doc.go

```diff
@@ -0,0 +1,117 @@
+
+	0. Define all possible in the package errors as public variables and let consumers compare returned errors to them.
+		For.Ex. https://golang.org/pkg/net/http/#pkg-variables ('Errors used by the HTTP server.' block)
+		+ very easy implementation
+		+ extremely performant
+		- consumer CAN redefine it
+		- such errors do not contain explicit information about particular case(s)
+	1. Define own Error type as structure with detail information fields.
+		For.Ex. https://golang.org/pkg/os/#PathError
+		+ strongly typed
+		+ as much information as fields provided
+		- consumer must know such Error structure
+		- data contract, rather than behaviour
+		- direct fields access = cosumer CAN change it
+	2. Define own Error interface (in combination with p.2) with contextual properties.
+		For.Ex. https://golang.org/pkg/net/#Error
+		+ strongly typed
+		+ behaviour contract, rather than data
+		Actually, this approach does not have irresolvable disadvantages, however:
+		- does not(at least easily) provide good Operational representation.
+		- not flexible enough to add new property on the fly.
+
+This package contains implementation of another approach which has its own
+advantages and, obviously, disadvantages. This solution declares single
+interface for all errors all over the code. It provides functionallity to describe
+execution flow and description contexts on creation of error. This could be done by
+passing `details` to ErrorFactory. Later this details are accessible via `DetailedError`
+interface:
+
+	* Detailed() string - returns Operational representation.
+	* Get(key interface{}) (interface{}, bool) - returns execution flow representation.
+
+It is also possible to combine both representations for one detail by passing Key and/or Value which conform(s)
+fmt.Stringer interface or is of type string.
+
+Thus, consumer of the following error
+
+	// Expose immutable error property
+	const ErrorIsCritical = myPackageErrorDetailStringer("Is Critical")
+
+	type myPackageErrorDetailStringer string
+	func (this myPackageErrorDetailStringer) String() string { return this }
+
+	func DoSomething() Error {
+
+		...
+
+		return errorFactory.New(
+			// Some error message
+			"Some error",
+			// The detail which extends dry GoLang error with context
+			struct{ Key, Value interface{}}{ Key: ErrorIsCritical })
+	}
+
+Is able to utilize it in any representation.
+
+	err := myPackage.DoSomething()
+	if err != nil {
+		// In this place Operations representation is utilized. Due to ErrorIsCritical
+		// conforms fmt.Stringer() err.Detailed() will log it.
+		log.Info(err.Detailed())
+
+		// Here Code representation is utilized.
+		// We define execution flow based on existance of myPackage.ErrorIsCritical flag in Error details.
+		if _, isCritical := Error.Get(myPackage.ErrorIsCritical); isCritical{
+			// Now we can manage CRITICAL error
+		} esle {
+			// Error is not critical
```

Comment:
1: what does it mean 'esle'?
2: we should have a chat about critical Errors. I don't see the point in them

Reply:
Sure. Seems like I have to bring better example.